waspls/src/Wasp/Backend/ConcreteParser/Internal.hs

```diff
@@ -0,0 +1,739 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+
+module Wasp.Backend.ConcreteParser.Internal
+  ( -- * Internal parsing library
+    GrammarRule,
+    parse,
+
+    -- * Combinators
+
+    -- | The fixity of the operators ('(<>)', '`as`', '(<$$>)', and '(<|>)') was
+    -- chosen to make using operators require as little parentheses as possible:
+    --
+    -- >>> Program <$$> lparen <> rparen
+    --       <|> string
+    -- Alternative (Group Program (Chain (Token LParen) (Token RParen))) (Token String)
+    as,
+    (<$$>),
+    perhaps,
+    (<|>),
+
+    -- * Primitive grammar rules
+
+    -- | These grammar rules (besides `eof`) consume a single token of the kind
+    -- matching their name. The syntax node they produce has an "S.Token" kind,
+    -- unless otherwise specified.
+    lparen,
+    rparen,
+    lsquare,
+    rsquare,
+    lcurly,
+    rcurly,
+    comma,
+    colon,
+    kwImport,
+    kwFrom,
+    kwTrue,
+    kwFalse,
+    string,
+    int,
+    double,
+    lquote,
+    rquote,
+    quoted,
+    -- | A primitive for "T.Identifier" is not included: this should always be
+    -- done with an `as`, i.e. @"T.Identifier" ``as`` "S.DictKey"@
+    eof,
+
+    -- * Error Recovery
+
+    -- | This parser has built-in automatic error recovery. Currently, error
+    -- recovery occurs when an error is produced inside of a group rule. Recovery
+    -- is based on the possible tokens that can follow the current rule.
+    --
+    -- At a basic level, we can describe the mechanism as follows:
+    --
+    -- For a "GrammarRule" @label '<$$>' first '<>' next@, if an error occurs
+    -- while parsing @first@, check if the token it errored on is in the follow
+    -- set for the group.
+    --
+    -- If yes, then there is a rule that can accept this token. So just continue
+    -- parsing. Otherwise, no rule has a use for the token, so discard it.
```

Comment:
Said this in another response too, but this section is meant to introduce the error recovery strategy in two parts. So I don't want to explain extended follow sets until unwinding is introduced.

In case you're worrying about the inconsistency causing some issues, there's the error recovery rules further on which is the full definition of how error recovery works.